core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -101,6 +101,53 @@ public static Snapshot oldestAncestor(Table table) {
     return ancestorsOf(start, lookup);
   }
 
+  /**
+   * Traverses the history of the table's current snapshot
+   * and finds the oldest ancestor snapshot after or equal to the timestamp in milliseconds.
+   *
+   * @param table           a table
+   * @param timestampMillis a timestamp in milliseconds
+   * @return the first snapshot from the given timestamp, or null if the current snapshot is older than the timestamp
+   * @throws IllegalStateException if the first snapshot from the given time can't be determined
+   */
+  public static Snapshot oldestAncestorFrom(Table table, Long timestampMillis) {
+    Snapshot currentSnapshot = table.currentSnapshot();
+    long timestamp = timestampMillis == null ? -1L : timestampMillis;
+    // If no snapshot exists or timestamp is higher than the current snapshot
+    if (currentSnapshot == null || currentSnapshot.timestampMillis() < timestamp) {
+      return null;
+    }
+
+    // Traverse to the oldest snapshot after or equal to the timestamp
```

Comment:
This looks like it is now

"The oldest snapshot that was not created after the timestamp" - which would be good in the docs ^ too

Like
```
Given [10, 11, 12, 13, 14]
oldestAncestor(x <=10) = null
oldestAncestor(11)  = 11
oldestAncestor(14) = 14
oldestAncestor(x > 14) = 14
```

Reply:
Ok I forgot how directions work. Disregard this

Reply:
So If I have this right the method is basically (assuming ordered by timestamp)

    StreamSupport.stream(currentAncestors(table).spliterator(), false)
        .filter(snapshot -> snapshot.timestampMillis() >= targetTimestamp) // For elements that are as new as or newer than the target 
        .reduce((first, second) -> second) // Return the last element we see (although a boolean could be put in here to select the oldest of first and second if ordering of currentAncestors is not guaranteed ... I think it is though)
        .orElse(null);

Reply:
Actually don't do that because that forces a full iteration through all snapshots ... oh well.

Reply:
Ok. Yes, it won't break early.

> Actually don't do that because that forces a full iteration through all snapshots ... oh well.